wasm.rkt

```diff
@@ -60,14 +63,92 @@
 ;; Delta
 
 (define-metafunction WASM-eval
-  delta : op v ... -> v
+  delta : op v ... -> v ∨ trap
+  ;;; BINOPS
   [(delta add (const c_0) (const c_1)) (const ,(+ (term c_0) (term c_1)))]
+  [(delta sub (const c_0) (const c_1)) (const ,(- (term c_0) (term c_1)))]
+  [(delta mul (const c_0) (const c_1)) (const ,(* (term c_0) (term c_1)))]
+  [(delta div (const c_0) (const c_1)) ,(if (= (term c_1) 0)
+                                            (term trap)
+                                            (term (const ,(/ (term c_0) (term c_1)))))]
+  [(delta rem (const c_0) (const c_1)) ,(if (= (term c_1) 0)
+                                            (term trap)
+                                            (term (const ,(remainder (term c_0) (term c_1)))))]
+  [(delta and (const c_0) (const c_1)) (const ,(bitwise-and (term c_0) (term c_1)))]
+  [(delta or (const c_0) (const c_1)) (const ,(bitwise-ior (term c_0) (term c_1)))]
+  [(delta xor (const c_0) (const c_1)) (const ,(bitwise-xor (term c_0) (term c_1)))]
+  ; Note that the shift and rotate ops only make sense in the context of a fixed bit width.
+  ; Here, we assume that all operands are 32 bit.
+  [(delta shl (const c_0) (const c_1))
+   (const ,(bitwise-and #xFFFFFFFF
+                        (arithmetic-shift (term c_0) (term c_1))))]
+  [(delta shr (const c_0) (const c_1))
+   (const ,(arithmetic-shift (term c_0) (- (term c_1))))]
+  [(delta rotl (const c_0) (const c_1))
+   (const ,(if (< (term c_1) 0)
+               (term (rotr (const c_0) (const ,(- (term c_1)))))
+               (bitwise-ior (bitwise-and #xFFFFFFFF
+                                         (arithmetic-shift (term c_0) (remainder (term c_1) 32)))
+                            (arithmetic-shift (term c_0)
+                                              (- (remainder (term c_1) 32) 32)))))]
+  [(delta rotr (const c_0) (const c_1))
+   (const ,(if (< (term c_1) 0)
+               (term (rotl (const c_0) (const ,(- (term c_1)))))
+               (bitwise-ior (arithmetic-shift (term c_0) (- (remainder (term c_1) 32)))
+                            (bitwise-and #xFFFFFFFF
+                                         (arithmetic-shift (term c_0)
+                                                           (- 32 (remainder (term c_1) 32)))))))]
+
+  ;;; UNOPS
+  ; These unops only make sense in the context of a fixed bit width. Here, 32 bits.
+  [(delta ctz (const c_in))
+   ,(local [(define (ctz-internal in rem)
+              (if (or (= 0 rem)
+                      (= 1 (bitwise-and 1 in)))
+                  0
+                  (+ (- (- (bitwise-and 1 in) 1))
```

Comment:
nit: i think this would be clearer if it read if `if (= (bitwise-and 1 in) 0) 1 0)`

Reply:
I do think this would make it considerably cleaner. Took me a few mins to understand the arithmetic and the if would have been clearer. But yeah no big deal either way.

Reply:
yeah I closed this because I addressed it (just didn't push yet) not because I disagree

Reply:
got it my bad